File: lib/features/authentication/test/authentication_test.dart
import 'package:flutter_test/flutter_test.dart';
import 'package:passdriver/features/authentication/authentication_provider.dart';
import 'package:mocktail/mocktail.dart';

class MockAuthenticationProvider extends Mock implements AuthenticationProvider {}

void main() {
  late MockAuthenticationProvider mockAuthenticationProvider;

  setUp(() {
    mockAuthenticationProvider = MockAuthenticationProvider();
  });

  test('should return error when authentication fails', () async {
    when(() => mockAuthenticationProvider.authenticate(any(), any())).thenThrow(Exception('Authentication failed'));
    expect(() async => await mockAuthenticationProvider.authenticate('username', 'password'), throwsException);
  });

  test('should return error when storage limit is reached', () async {
    when(() => mockAuthenticationProvider.checkStorageLimit()).thenThrow(Exception('Storage limit reached'));
    expect(() async => await mockAuthenticationProvider.checkStorageLimit(), throwsException);
  });
}

File: backend/src/hono/routes/authentication/test/authentication.test.ts
import { Hono } from 'hono';
import { authenticationRoute } from '../authentication';
import { mockRequest } from '../../../../test/helpers';

describe('Authentication route', () => {
  it('should return 401 when authentication fails', async () => {
    const c = mockRequest({ username: 'username', password: 'wrong_password' });
    const response = await authenticationRoute(c);
    expect(response.status).toBe(401);
  });

  it('should return 507 when storage limit is reached', async () => {
    const c = mockRequest({ username: 'username', password: 'password' });
    c.env.KV.get = async () => 'true';
    const response = await authenticationRoute(c);
    expect(response.status).toBe(507);
  });
});
